x/valset-pref/validator_set.go

```diff
@@ -96,28 +137,36 @@ func (k Keeper) UndelegateFromValidatorSet(ctx sdk.Context, delegatorAddr string
 	// the total amount the user wants to undelegate
 	tokenAmt := sdk.NewDec(coin.Amount.Int64())
 
-	totalAmountFromWeights := sdk.NewDec(0)
-	for _, val := range existingSet.Preferences {
-		totalAmountFromWeights = totalAmountFromWeights.Add(val.Weight.Mul(tokenAmt))
-	}
-
-	// Handles rounding for ex: 9999.99999 = 10000.0
-	totalAmountFromWeights = totalAmountFromWeights.RoundInt().ToDec()
-
-	if !totalAmountFromWeights.Equal(tokenAmt) {
-		return fmt.Errorf("The undelegate total do not add up with the amount calculated from weights expected %s got %s", tokenAmt, totalAmountFromWeights)
+	err = k.CheckUndelegateTotalAmount(tokenAmt, existingSet.Preferences)
+	if err != nil {
+		return err
 	}
 
-	for _, val := range existingSet.Preferences {
-		// Calculate the amount to undelegate based on the existing weights
-		amountToUnDelegate := val.Weight.Mul(tokenAmt)
+	// totalDelAmt is the amount that keeps running track of the amount of tokens undelegated
+	totalUnDelAmt := sdk.NewInt(0)
+	// tokenAmt is the current amount to undelegate
+	amountToUnDelegate := sdk.NewInt(0)
+	for i, val := range existingSet.Preferences {
 
 		valAddr, validator, err := k.getValAddrAndVal(ctx, val.ValOperAddress)
 		if err != nil {
 			return err
 		}
 
-		sharesAmt, err := validator.SharesFromTokens(amountToUnDelegate.TruncateInt())
+		// in the last valset iteration we dont calculate it from shares using decimals and trucation,
+		// we use whats remaining to get more accurate value
+		if len(existingSet.Preferences)-1 == i {
+			amountToUnDelegate = coin.Amount.Sub(totalUnDelAmt).ToDec().TruncateInt()
+		} else {
+			// Calculate the amount to undelegate based on the existing weights
+			amountToUnDelegate = val.Weight.Mul(tokenAmt).TruncateInt()
+			totalUnDelAmt = totalUnDelAmt.Add(amountToUnDelegate)
+		}
+
```

Comment:
test case for this needs to be added

Reply:
yup added 2 that touches this logic (https://github.com/osmosis-labs/osmosis/pull/4143/commits/5c0b35f4e18968470c875431083a988e5a4e0a77#diff-e8d94c443b26e1cda354d150b30ce5542344099e0bbbb8e84e507d9df791daf7R359-R375)